### PR TITLE
Auto-balance tests, add test reporters

### PR DESCRIPTION
**Pull request by frenchy64**
Closes https://github.com/threatgrid/iroh/issues/4378

Summary: 40 second faster build with auto test rebalancing.

Test results INCLUDING QUEUING TIME! See below for detailed spits:
| Old build (9 splits) | New build (9 splits) | New build (10 splits) 
| --- | --- | ---- |
| 8m 58s | 8m 57s | 8m 19s (40s faster) |

Split range:
| Old build (9 splits) | New build (9 splits) | New build (10 splits) 
| --- | --- | ---- |
| 5m-8m | 5m-7m | 5m-7m |

Old PR build (9 splits): https://github.com/threatgrid/ctia/runs/1391155782

```
GitHub Actions Pull Request / shellcheck (pull_request) Successful in 8s
GitHub Actions Pull Request / test-matrix (0) (pull_request) Successful in 7m
GitHub Actions Pull Request / test-matrix (1) (pull_request) Successful in 7m
GitHub Actions Pull Request / test-matrix (2) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (3) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (4) (pull_request) Successful in 7m
GitHub Actions Pull Request / test-matrix (5) (pull_request) Successful in 8m
GitHub Actions Pull Request / test-matrix (6) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (7) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (8) (pull_request) Successful in 8m
11 completed jobs 
in 8m 58s (INCLUDING queuing time!)
```

New build BEFORE AUTO REBALANCE (9 splits): https://github.com/threatgrid/ctia/actions/runs/359977570
```
GitHub Actions Pull Request / setup (pull_request) Successful in 12s
GitHub Actions Pull Request / test-matrix (0) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (1) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (2) (pull_request) Successful in 8m
GitHub Actions Pull Request / test-matrix (3) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (4) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (5) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (6) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (7) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (8) (pull_request) Successful in 6m
GitHub Actions Pull Request / all-pr-checks (pull_request) Successful in 11s
11 completed jobs 
in 9m 29s  (INCLUDING queuing time!)
```

New build AFTER AUTO REBALANCE (9 splits): https://github.com/threatgrid/ctia/actions/runs/359996637
```
GitHub Actions Pull Request / setup (pull_request) Successful in 13s
GitHub Actions Pull Request / test-matrix (0) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (1) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (2) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (3) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (4) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (5) (pull_request) Successful in 7m
GitHub Actions Pull Request / test-matrix (6) (pull_request) Successful in 7m
GitHub Actions Pull Request / test-matrix (7) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (8) (pull_request) Successful in 6m
GitHub Actions Pull Request / all-pr-checks (pull_request) Successful in 11s
11 completed jobs 
in 8m 57s  (INCLUDING queuing time!)
```

New build AFTER AUTO REBALANCE (10 splits): https://github.com/threatgrid/ctia/actions/runs/360068378
```
GitHub Actions Pull Request / setup (pull_request) Successful in 9s
GitHub Actions Pull Request / test-matrix (0) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (1) (pull_request) Successful in 7m
GitHub Actions Pull Request / test-matrix (2) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (3) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (4) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (5) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (6) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (7) (pull_request) Successful in 5m
GitHub Actions Pull Request / test-matrix (8) (pull_request) Successful in 6m
GitHub Actions Pull Request / test-matrix (9) (pull_request) Successful in 5m
GitHub Actions Pull Request / all-pr-checks (pull_request) Successful in 9s
12 completed jobs 
in 8m 19s (INCLUDING queuing time!)
```

This PR adds a fully-automated test balancing strategy based on prior test timings. The build cache is used to store previous timings for the same pull-request, or the latest timings globally.

For local experimentation to help find bottlenecks etc.,, instructions are added to fetch these timings locally.

You can see it in action by scrolling to the bottom of this PR and looking at the build. As of writing, all jobs were 4-6 minutes each (click `Show all checks`).

The basic algorithm takes as input timings of previous test runs, and iterates over the test namespaces slowest-to-fastest allocating them to the currently-fastest job. This is an approximate algorithm since the general problem of [scheduling](https://en.wikipedia.org/wiki/Multiprocessor_scheduling) is NP-hard (ie., exponential). With 10 splits, theoretically less than 1 second is "wasted" (ie., the predicted difference between the slowest and fastest job is less than 1 second). In practice, it's usually within a minute.

AFAICT 14 splits would be more optimal, because longer than that would create large bottlenecks around the longest running namespaces (migration tests IIRC). Unfortunately, we have a 20 build limit before queuing, so I chose 10 so we can build 2 commits in the same time as building 1.

I'm not sure how to replicate the same setup in Travis, so we're using the "slow-namespace" heuristic there (for push builds).

Assertions are used to ensure all scheduled namespaces are run. An interesting case is where we've recorded a previous timing for a namespace that doesn't exist anymore. We should discard it in this case, and not run tests for the missing namespace (checked via assertion).

To verify the latest timings are being used:
1. Find 2 builds that occurred one after another
2. In the earlier build, in the `all-pr-checks` job, note the cache id saved in the `Timing results Cache` step. Note that in `Post Timing results Cache` the cache is indeed saved.
  - eg., `ctia-timings-refs/pull/1028/merge-e937c41ae91edbea4f07afec0a8c44aa69407f76`
  - [example build](https://github.com/threatgrid/ctia/runs/1365819068#step:6:14)
3. In the later build, in the `setup` job, the `Timing Cache` step should have the same cache ID:
  - `Cache restored from key: ctia-timings-refs/pull/1028/merge-e937c41ae91edbea4f07afec0a8c44aa69407f76`
  - [example build](https://github.com/threatgrid/ctia/runs/1365821108?check_suite_focus=true#step:7:21)

<a name="qa">[§](#qa)</a> QA
============================

No QA is needed.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: Auto-balance tests, add test reporters
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -19,36 +19,34 @@ env:
...
     runs-on: ubuntu-18.04
     strategy:
       matrix:
-        this-split: [0,1,2,3,4,5,6,7,8] # (range ${CTIA_NSPLITS})
+        this-split: [0,1,2,3,4,5,6,7,8,9] # (range ${CTIA_NSPLITS})
     env:
       CTIA_THIS_SPLIT: ${{ matrix.this-split }}
     steps:
       - uses: actions/checkout@v2
+      - name: ES5 setup
```
Comment (frenchy64) [inline .github/workflows/pr.yml:38]: Moving docker start as early as possible in the build seems a bit faster.

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -19,36 +19,34 @@ env:
...
-          (set -x && cp "shellcheck-${SHELLCHECK_VERSION}/shellcheck" "${BIN_DIR}/shellcheck")
-          (set -x && $SHELLCHECK_COMMAND --version)
-      - name: Run shellcheck
-        run: ./scripts/shellcheck-build.sh
   test-matrix:
     runs-on: ubuntu-18.04
     strategy:
       matrix:
-        this-split: [0,1,2,3,4,5,6,7,8] # (range ${CTIA_NSPLITS})
+        this-split: [0,1,2,3,4,5,6,7,8,9] # (range ${CTIA_NSPLITS})
```
Comment (frenchy64) [inline .github/workflows/pr.yml:33]: Moved shellcheck to each job, using the extra job for tests.

**Review by vemv — DISMISSED** [review]
I'll leave some feedback over chat

File: .github/workflows/pr.yml (hunk 1 above)
Comment (ereteog) [reply .github/workflows/pr.yml:38]: :-)
```suggestion
      - name: ES setup
```

File: dev/ctia/dev/split_tests.clj [hunk 3]
```diff
@@ -1,7 +1,10 @@
 (ns ctia.dev.split-tests
   (:require [circleci.test :as t]
+            [clojure.data.priority-map :refer [priority-map-keyfn]]
```
Comment (ereteog) [inline dev/ctia/dev/split_tests.clj:3]: just discovered these collections.
I like it :-)

**Review by ereteog — APPROVED** [review]
LGTM!

File: dev/ctia/dev/split_tests.clj [hunk 4]
```diff
@@ -98,6 +101,74 @@
...
+                                   (map vector
+                                        extra-namespaces
+                                        ;; allocated last
+                                        (repeat {:elapsed-ns 0})))
+                           (map (fn [[nsym {:keys [elapsed-ns]}]]
+                                  {:pre [(simple-symbol? nsym)
+                                         (<= 0 elapsed-ns)]}
+                                  [elapsed-ns nsym]))
+                           (into (sorted-set))
+                           rseq)]
```
[the remark this answers is not in the document]
Comment (frenchy64) [reply dev/ctia/dev/split_tests.clj:130]: Oh, I guess this is just a `reduce` right? The priority map is the accumulator and the namespace info is the collection.

File: dev/ctia/dev/split_tests.clj (hunk 3 above)
Comment (frenchy64) [reply dev/ctia/dev/split_tests.clj:3]: Yes this saved my butt.

**Comment by frenchy64** [comment]
@vemv let's discuss your concerns in https://github.com/threatgrid/iroh/issues/4378, which I just created.

File: dev/ctia/dev/split_tests.clj [hunk 5]
```diff
@@ -98,6 +101,69 @@
...
+                                                     (repeat {:elapsed-ns 0})))
+                                        (map (fn [[nsym {:keys [elapsed-ns]}]]
+                                               {:pre [(simple-symbol? nsym)
+                                                      (<= 0 elapsed-ns)]}
+                                               [elapsed-ns nsym]))
+                                        (into (sorted-set))
+                                        rseq)
+        ;; algorithm: always allocate new work to the fastest job, and process
+        ;;            work from slowest to fastest
+        splits (reduce (fn [so-far [elapsed-ns nsym :as current-timing]]
```
Comment (ereteog) [inline dev/ctia/dev/split_tests.clj:130]: this `reduce` is much easier to read than the previous `loop`. Here we properly identify that we iterate on the ascending timings and accumulate the splits.
thx.

Comment (frenchy64) [reply dev/ctia/dev/split_tests.clj:130]: Excellent thanks!

File: .github/workflows/pr.yml [hunk 6]
```diff
@@ -3,95 +3,274 @@ name: GitHub Actions Pull Request
...
-  CTIA_HOOK_ES_PORT: 9200
-  CTIA_HOOK_ES_INDEXNAME: elasticsearch
-  CTIA_HOOK_REDIS_URI: redis://127.0.0.1:6379
   # number of test-matrix builds
-  # (9 jobs for test-matrix + 1 for shellcheck) + 1 for fan-in
+  # 1 setup + 10 test jobs + 1 for fan-in
   # allows 2 commits to build simultaneously, since one build
   # will take at most 10 builds at any one time
-  CTIA_NSPLITS: 9
+  CTIA_NSPLITS: 10
```
Comment (ereteog) [inline .github/workflows/pr.yml:16]: I would keep the same number of splits to compare the build time with previous solution.

**Comment by frenchy64** [comment]
@vemv merging this before I can get your feedback, more context here https://github.com/threatgrid/iroh/issues/4378#issuecomment-726321874

**Pull request merged**